### PR TITLE
Support Variable Fonts in Desktop, Native, Web

### PR DESCRIPTION
**Pull request by adamglin0**
Support for **Variable Font** in platforms other than Android.

Fix Issue [Youtrack](https://youtrack.jetbrains.com/issue/CMP-3127/Add-support-for-variable-fonts-on-iOS-desktop)

## Testing

Write Sample In Compose MPP Demo

- Desktop run `./gradlew :compose:mpp:demo:runDesktop`
- JS run `./gradlew :compose:mpp:demo:jsRun`
- IOS run `./compose/mpp/demo/regenerate_xcode_project.sh`

> For more details on how to run the mpp demo, please refer to MULTIPLATFORM.md
> Should tested by QA.

## How to use

```kotlin
Font(
      identity = "FontName",
      data = fontData,
      variationSettings = FontVariation.Settings()
)
```
## Record

https://github.com/user-attachments/assets/8f106402-c3e9-470d-9eec-db02590cf1f3

https://github.com/user-attachments/assets/853f248a-fad3-4549-aa7d-c3b714ac2ccd

https://github.com/user-attachments/assets/bb6f3a83-8a00-4e82-87fb-5871f3e6ea55

### Features - Multiple Platforms
- Support Variable Fonts In All Platforms

### CODE REVIEW

**Comment by adamglin0** [comment]
@MatkovIvan  Hello, how do I add a reviewer to this PR?

File: compose/mpp/demo/build.gradle.kts [hunk 1]
```diff
@@ -21,6 +21,9 @@ import org.jetbrains.kotlin.gradle.plugin.mpp.KotlinNativeTarget
 import org.jetbrains.kotlin.gradle.plugin.mpp.NativeBuildType
 import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
 import org.jetbrains.kotlin.gradle.targets.js.webpack.KotlinWebpackConfig
+import org.jetbrains.kotlin.gradle.tasks.FatFrameworkTask
+import org.jetbrains.kotlin.gradle.tasks.KotlinNativeLink
+import org.jetbrains.kotlin.konan.target.CompilerOutputKind
```
Comment (terrakok) [inline compose/mpp/demo/build.gradle.kts]: added by mistake?

File: compose/ui/ui-text/src/desktopMain/kotlin/androidx/compose/ui/text/platform/DesktopFont.desktop.kt [hunk 2]
```diff
@@ -180,7 +190,7 @@ private fun typefaceResource(resourceName: String): SkTypeface {
         ?: (::typefaceResource.javaClass).getResourceAsStream(resourceName)
         ?: error("Can't load font from $resourceName")
 
-    val bytes = resource.use { it.readAllBytes() }
+    val bytes = resource.use { it.readBytes() }
```
Comment (terrakok) [inline compose/ui/ui-text/src/desktopMain/kotlin/androidx/compose/ui/text/platform/DesktopFont.desktop.kt]: Why was it changed? `readBytes` is able to read only part of the data

File: compose/ui/ui-text/src/desktopMain/kotlin/androidx/compose/ui/text/platform/DesktopFont.desktop.kt [hunk 3]
```diff
@@ -47,7 +48,8 @@ actual sealed class PlatformFont : Font {
 class ResourceFont internal constructor(
     val name: String,
     override val weight: FontWeight = FontWeight.Normal,
-    override val style: FontStyle = FontStyle.Normal
+    override val style: FontStyle = FontStyle.Normal,
+    override val variationSettings: FontVariation.Settings = FontVariation.Settings(weight, style),
```
Comment (terrakok) [inline compose/ui/ui-text/src/desktopMain/kotlin/androidx/compose/ui/text/platform/DesktopFont.desktop.kt:52]: Just curios, what will happen if I use different a weight and a style here?
`Font("name", weight1, style1, FontVariation.Settings(weight2, style2))`

File: compose/ui/ui-text/src/skikoMain/kotlin/androidx/compose/ui/text/platform/PlatformFont.skiko.kt [hunk 4]
```diff
@@ -120,8 +127,16 @@ fun Font(
...
+    style: FontStyle = FontStyle.Normal,
+): Font = LoadedFont(identity, getData, weight, style, FontVariation.Settings())
+
+fun Font(
+    identity: String,
+    getData: () -> ByteArray,
+    weight: FontWeight = FontWeight.Normal,
+    style: FontStyle = FontStyle.Normal,
+    variationSettings: FontVariation.Settings = FontVariation.Settings(weight, style),
+): Font = LoadedFont(identity, getData, weight, style, variationSettings)
```
Comment (terrakok) [inline compose/ui/ui-text/src/skikoMain/kotlin/androidx/compose/ui/text/platform/PlatformFont.skiko.kt]: Could you explain why we need one more method?

File: compose/ui/ui-text/src/skikoMain/kotlin/androidx/compose/ui/text/platform/PlatformFont.skiko.kt [hunk 5]
```diff
@@ -147,14 +162,28 @@ fun Font(
...
+    weight: FontWeight = FontWeight.Normal,
+    style: FontStyle = FontStyle.Normal,
+    variationSettings: FontVariation.Settings = FontVariation.Settings(weight, style),
+): Font = Font(
+    identity = identity,
+    getData = { data },
+    weight = weight,
+    style = style,
+    variationSettings = variationSettings,
+)
```
Comment (terrakok) [inline compose/ui/ui-text/src/skikoMain/kotlin/androidx/compose/ui/text/platform/PlatformFont.skiko.kt]: and here as well

File: compose/ui/ui-text/src/skikoMain/kotlin/androidx/compose/ui/text/platform/PlatformFont.skiko.kt [hunk 6]
```diff
@@ -320,3 +351,18 @@ private val GenericFontFamiliesMapping: Map<String, List<String>> by lazy {
...
+internal fun FontVariation.Settings.toSikaFontVariationList(): List<org.jetbrains.skia.FontVariation> {
+    return settings.map { setting ->
+        org.jetbrains.skia.FontVariation(setting.axisName, setting.toVariationValue(null))
+    }
+}
+
+/**
+ * Bind the font variation settings to the Skia typeface.
+ */
+internal fun SkTypeface.bindVariantSettings(variationSettings: FontVariation.Settings): SkTypeface {
```
Comment (terrakok) [inline compose/ui/ui-text/src/skikoMain/kotlin/androidx/compose/ui/text/platform/PlatformFont.skiko.kt]: Since it clones the original font, I would name it `cloneWithVariationSettings`

**Review by terrakok — CHANGES_REQUESTED** [review]

File: compose/ui/ui-text/src/desktopMain/kotlin/androidx/compose/ui/text/platform/DesktopFont.desktop.kt (hunk 3 above)
Comment (adamglin0) [reply compose/ui/ui-text/src/desktopMain/kotlin/androidx/compose/ui/text/platform/DesktopFont.desktop.kt:52]: If variationSettings is set, it will follow the settings of variationSettings.

File: compose/ui/ui-text/src/skikoMain/kotlin/androidx/compose/ui/text/platform/PlatformFont.skiko.kt (hunk 4 above)
Comment (adamglin0) [reply compose/ui/ui-text/src/skikoMain/kotlin/androidx/compose/ui/text/platform/PlatformFont.skiko.kt]: Sorry, I referred to the implementation on Android but didn't think it through carefully. 🙇

Because there are two methods on Android, the difference lies in whether there are variationSettings.
``` kotlin
@Stable
fun Font(
    resId: Int,
    weight: FontWeight = FontWeight.Normal,
    style: FontStyle = FontStyle.Normal,
    loadingStrategy: FontLoadingStrategy = FontLoadingStrategy.Blocking
): Font = ResourceFont(resId, weight, style, FontVariation.Settings(), loadingStrategy)

@ExperimentalTextApi
fun Font(
    resId: Int,
    weight: FontWeight = FontWeight.Normal,
    style: FontStyle = FontStyle.Normal,
    loadingStrategy: FontLoadingStrategy = FontLoadingStrategy.Blocking,
    variationSettings: FontVariation.Settings = FontVariation.Settings(weight, style)
): Font = ResourceFont(resId, weight, style, variationSettings, loadingStrategy)
```
Looking at it now, methods with variationSettings should replace methods without variationSettings in the future.

Users of variable fonts will only choose the second method, while for regular fonts, users can choose either method. This is because if it is not a variable font, the settings for variationSettings will be omitted, which is also fine.

So whether the user is using regular fonts or variable fonts, the methods with variationSettings will work normally.

Therefore, I will only keep the methods with variationSettings. Thank you.

File: compose/ui/ui-text/src/desktopMain/kotlin/androidx/compose/ui/text/platform/DesktopFont.desktop.kt (hunk 3 above)
Comment (adamglin0) [reply compose/ui/ui-text/src/desktopMain/kotlin/androidx/compose/ui/text/platform/DesktopFont.desktop.kt:52]: In summary:

1. Non-variable fonts define the font style only according to weight and style.

2. Variable fonts define the font style only according to variationSettings.

3. Because the default values of variationSettings are derived from style and style, in the case of variable fonts with weight, style, and variationSettings set, the font style will be determined by variationSettings.

File: compose/ui/ui-text/src/skikoMain/kotlin/androidx/compose/ui/text/platform/PlatformFont.skiko.kt (hunk 6 above)
Comment (adamglin0) [reply compose/ui/ui-text/src/skikoMain/kotlin/androidx/compose/ui/text/platform/PlatformFont.skiko.kt]: Updated function name and comment

**Review by terrakok — APPROVED** [review]

File: compose/ui/ui-text/src/skikoMain/kotlin/androidx/compose/ui/text/platform/PlatformFont.skiko.kt (hunk 4 above)
Comment (terrakok) [reply compose/ui/ui-text/src/skikoMain/kotlin/androidx/compose/ui/text/platform/PlatformFont.skiko.kt]: As I see, the difference here is when variation settings are empty then the clone function doesn't make a clone of the font

File: compose/ui/ui-text/src/skikoMain/kotlin/androidx/compose/ui/text/platform/PlatformFont.skiko.kt [hunk 7]
```diff
@@ -282,10 +317,15 @@ private val GenericFontFamiliesMapping: Map<String, List<String>> by lazy {
             mapOf(
                 FontFamily.SansSerif.name to listOf("Noto Sans", "DejaVu Sans", "Arial"),
                 FontFamily.Serif.name to listOf("Noto Serif", "DejaVu Serif", "Times New Roman"),
-                FontFamily.Monospace.name to listOf("Noto Sans Mono", "DejaVu Sans Mono", "Consolas"),
+                FontFamily.Monospace.name to listOf(
+                    "Noto Sans Mono",
+                    "DejaVu Sans Mono",
+                    "Consolas"
+                ),
```
Comment (kropp) [inline compose/ui/ui-text/src/skikoMain/kotlin/androidx/compose/ui/text/platform/PlatformFont.skiko.kt]: Here and in other places in this file: please avoid unrelated changes.
They make the PR longer and harder to review and are potential source of future merge conflicts.

File: compose/ui/ui-text/src/skikoMain/kotlin/androidx/compose/ui/text/platform/PlatformFont.skiko.kt [hunk 8]
```diff
@@ -320,3 +375,21 @@ private val GenericFontFamiliesMapping: Map<String, List<String>> by lazy {
             )
     }
 }
+
+internal fun FontVariation.Settings.toSikaFontVariationList(): List<org.jetbrains.skia.FontVariation> {
```
Comment (kropp) [inline compose/ui/ui-text/src/skikoMain/kotlin/androidx/compose/ui/text/platform/PlatformFont.skiko.kt]: typo: Sika → Skia

**Review by kropp — CHANGES_REQUESTED** [review]

File: compose/ui/ui-text/src/skikoMain/kotlin/androidx/compose/ui/text/platform/PlatformFont.skiko.kt [hunk 9]
```diff
@@ -295,22 +331,37 @@ private val GenericFontFamiliesMapping: Map<String, List<String>> by lazy {
...
+                FontFamily.Serif.name to listOf(
+                    ".AppleSystemUIFontSerif",
+                    "Times",
+                    "Times New Roman"
+                ),
+                FontFamily.Monospace.name to listOf(
+                    ".AppleSystemUIFontMonospaced",
+                    "Menlo",
+                    "Courier"
+                ),
```
Comment (kropp) [inline compose/ui/ui-text/src/skikoMain/kotlin/androidx/compose/ui/text/platform/PlatformFont.skiko.kt]: Please revert all unnecessary changes in this file

**Review by kropp — CHANGES_REQUESTED** [review]

File: compose/ui/ui-text/src/skikoMain/kotlin/androidx/compose/ui/text/platform/PlatformFont.skiko.kt [hunk 10]
```diff
@@ -110,18 +116,31 @@ class LoadedFont internal constructor(
...
  * @see FontFamily
  */
 fun Font(
     identity: String,
     getData: () -> ByteArray,
     weight: FontWeight = FontWeight.Normal,
-    style: FontStyle = FontStyle.Normal
-): Font = LoadedFont(identity, getData, weight, style)
+    style: FontStyle = FontStyle.Normal,
+    variationSettings: FontVariation.Settings = FontVariation.Settings(weight, style),
```
Comment (MatkovIvan) [inline compose/ui/ui-text/src/skikoMain/kotlin/androidx/compose/ui/text/platform/PlatformFont.skiko.kt]: It breaks binary compatibility for public/stable method. You need to keep the previous method as-is or under `@Deprecated(HIDDEN)`

File: compose/ui/ui-text/src/skikoMain/kotlin/androidx/compose/ui/text/platform/PlatformFont.skiko.kt [hunk 11]
```diff
@@ -137,22 +156,36 @@ private class SkiaBackedTypeface(
...
  *
  * @see FontFamily
  */
 fun Font(
     identity: String,
     data: ByteArray,
     weight: FontWeight = FontWeight.Normal,
-    style: FontStyle = FontStyle.Normal
+    style: FontStyle = FontStyle.Normal,
+    variationSettings: FontVariation.Settings = FontVariation.Settings(weight, style),
```
Comment (MatkovIvan) [inline compose/ui/ui-text/src/skikoMain/kotlin/androidx/compose/ui/text/platform/PlatformFont.skiko.kt]: It breaks binary compatibility for public/stable method. You need to keep the previous method as-is or under `@Deprecated(HIDDEN)`

**Review by MatkovIvan — DISMISSED** [review]

File: compose/ui/ui-text/src/skikoMain/kotlin/androidx/compose/ui/text/platform/PlatformFont.skiko.kt (hunk 10 above)
Comment (adamglin0) [reply compose/ui/ui-text/src/skikoMain/kotlin/androidx/compose/ui/text/platform/PlatformFont.skiko.kt]: I will retain the previous method and then add a new method.

Comment (adamglin0) [reply compose/ui/ui-text/src/skikoMain/kotlin/androidx/compose/ui/text/platform/PlatformFont.skiko.kt]: @MatkovIvan I have a question: will adding parameters to class constructors also break binary compatibility? What should I do?🙇

Comment (MatkovIvan) [reply compose/ui/ui-text/src/skikoMain/kotlin/androidx/compose/ui/text/platform/PlatformFont.skiko.kt]: Only if it's a part of the constructor. Just add another secondary constructor with the old signature + hidden deprecation

**Review by kropp — APPROVED** [review]

**Comment by skaldebane** [comment]
Hi there! Been trying to use this in `1.8.0-alpha01`, but it seems that there's no easy API to do so.

Do I have to take the binary font data, then pass it to the `Font` function that takes an `identity` and `data`? That's pretty cumbersome compared to the Android approach.

Would love to have this as an extra `variationSettings` parameter on the Composable `Font` function. That way it'll be much closer to the Android approach and much more ergonomic to use.

**Comment by adamglin0** [comment]
> Hi there! Been trying to use this in `1.8.0-alpha01`, but it seems that there's no easy API to do so.
> 
> Do I have to take the binary font data, then pass it to the `Font` function that takes an `identity` and `data`? That's pretty cumbersome compared to the Android approach.
> 
> Would love to have this as an extra `variationSettings` parameter on the Composable `Font` function. That way it'll be much closer to the Android approach and much more ergonomic to use.

Do you mean org.jetbrains.compose.resources.Font()?
https://youtrack.jetbrains.com/issue/CMP-7088/Support-FontVariation.Settings-in-compose.component.resource
https://github.com/JetBrains/compose-multiplatform/pull/5182

**Comment by terrakok** [comment]
https://github.com/JetBrains/compose-multiplatform/pull/5183
